### PR TITLE
LobbyInfo

### PR DESCRIPTION
**Pull request by Bloodiko**
Adds Lobby details to RPC

Queue Name
Current Player Count
Max Player Count

curious if we manage to add a Join Lobby button. kekw - should be possible with the api maybe.

### CODE REVIEW

File: league_rpc_linux/lcu/LCU.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    League Client Update (LCU) API class.
+    """
+
+    def __init__(self):
+        self.lcu_connect = None
+        self.lcu_connect_port = None
+        self.lcu_connect_password = None
+        self.lcu_connect_protocol = "https"
+        self.lcu_connect_host = "127.0.0.1"
+        self.get_lcu_connect_details(proc_name="CrBrowserMain")
```
Comment (Its-Haze) [inline league_rpc_linux/lcu/LCU.py:16]: This is too generic, and won't work on windows. Opt into using lcu-driver instead.

My PR on that project makes it possible to find LeagueClientUx.exe, even if the process name is CrBrowserMain

File: league_rpc_linux/lcu/LCU.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        return self.get_lcu_connect_details(
+            proc_name="CrBrowserMain"
+        ) | self.get_lcu_connect_details(proc_name="LeagueClientUx.exe")
+
+    def getDataFromPath(self, path: str) -> dict:
+        """
+        Get data from path.
+        """
+        url = f"https://riot:{self.lcu_connect_password}@{self.lcu_connect_host}:{self.lcu_connect_port}{path}"
```
Comment (Its-Haze) [inline league_rpc_linux/lcu/LCU.py:47]: Even though this might work, I think it would be better if we sent in the riot:password in the Authorization header of the request.

Comment (Its-Haze) [reply league_rpc_linux/lcu/LCU.py:47]: But since i believe we should try to use lcu-driver (unless we have good reasons not to) we won't have to do these things to connect to lcu-api.

**Comment by Its-Haze** [comment]
If we are going to add additional features to the "in client" case, Then we should atleast have the following in place

1. large_image should be reflected based on what type of lobby the user is in.. OR it should be set to the summoners icon of the user.
2. We should try to use lcu-driver if we can (but if we have good reasons not to, then we can try to do things on our own)
3. The RPC should be very similar to how windows version does it, BUT we should make our slightly better. (Meaning that whatever they have, we should also have but slightly more)
4. If possible, We could try to use websockets instead of http to let the RPC only change when the user actually does a change.
I believe that if we should do this, and use lcu-driver. then that process needs to be in it's own thread, and have it's own queue of responses from the (wss)
5. Having a "Join lobby" button on discord would be nice to have but it should be optional.
5. Customizability: We should let users decide what information they want to show, and what to keep private. (Maybe they don't want to have a button asking people to join) or they just don't want people to know that they are in queue of a ranked game.

**Comment by Bloodiko** [comment]
Should the pypresence update also happen inside the separate thread ? - or keeping updating to main thread only, passing data back to main when requested in the 10 sec interval

File: league_rpc_linux/lcu/LCU.py (hunk 2 above)
Comment (Bloodiko) [reply league_rpc_linux/lcu/LCU.py]: Updated to lcu-driver

File: league_rpc_linux/lcu/LCU.py (hunk 1 above)
Comment (Bloodiko) [reply league_rpc_linux/lcu/LCU.py]: invalid - updated to lcu-driver

**Comment by Bloodiko** [comment]
i have adjusted most of the points. 

Adjustability would be for a later update - I guess this should be configureable inside a config file, as not to have 1 million config flags for each option ?

File: .gitignore [hunk 3]
```diff
@@ -21,3 +21,6 @@ instance/
 .coverage
 htmlcov/
 *.spec
+league-rpc-env/
+
+getConnectDetails.py
```
Comment (Its-Haze) [inline .gitignore:26]: What is this file, and why do we need to ignore it?

File: league_rpc_linux/lobby.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+from league_rpc_linux.lcu.LCU import getLcuAPI
```
Comment (Its-Haze) [inline league_rpc_linux/lobby.py:1]: Where is this located, can't see it in your PR.

**Review by Its-Haze — CHANGES_REQUESTED** [review]
Please add all files that are supposed to be a part of the pr.

File: league_rpc_linux/__main__.py [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import argparse
 import sys
 import time
+from multiprocessing import Process
+import nest_asyncio
```
Comment (Its-Haze) [inline league_rpc_linux/__main__.py:5]: This is pretty cool.. this is what i got stuck at, constantly got "RuntimeError: This event loop is already running" Cool that you found this!

File: league_rpc_linux/__main__.py [hunk 6]
```diff
@@ -1,6 +1,8 @@
 import argparse
 import sys
 import time
+from multiprocessing import Process
```
Comment (Its-Haze) [inline league_rpc_linux/__main__.py:4]: What's the reason for picking multiprocessing here instead of threading

File: league_rpc_linux/processes/LCU_Thread.py
Comment (Its-Haze) [inline league_rpc_linux/processes/LCU_Thread.py:1]: Please use lower cased characters on the python modules.

File: league_rpc_linux/__main__.py [hunk 7]
```diff
@@ -13,13 +15,19 @@
     check_league_client_process,
     player_state,
 )
+
 from league_rpc_linux.reconnect import discord_reconnect_attempt
+from league_rpc_linux.processes import LCU_Thread
```
Comment (Its-Haze) [inline league_rpc_linux/__main__.py:20]: Why import an entire module? would it not make more sence to import a function/class that we then call somewhere in our main loop?

If you import the entire module here, i believe it will execute lcu-driver before we even run the main loop.. which could be problematic

File: .gitignore (hunk 3 above)
Comment (Bloodiko) [reply .gitignore:26]: league-rpc-env is the default name for the python env according to your readme

getConnectDetails.py is a small script grabbing the LCU connect details from the process and prints them
This allowes me to easily copy them over to Insomnia for local testing of the endpoints
, not related to the Project, but only locally for my dev requirements. 
And its just a .gitignore anyway.  - If it matters, i'll remove it from the file.

File: league_rpc_linux/__main__.py [hunk 8]
```diff
@@ -35,6 +43,10 @@ def main(cli_args: argparse.Namespace):
         client_id=cli_args.client_id,
         wait_for_discord=cli_args.wait_for_discord,
     )
+
+    p = Process(target=LCU_Thread.start_connector, args=(rpc,))
```
Comment (Its-Haze) [inline league_rpc_linux/__main__.py:47]: Could you please explain what this will do exactly? and why args=(rpc,) is needed.

Maybe placing a comment above, giving a short summary of why this is needed would be good for future reference.

File: league_rpc_linux/lobby.py (hunk 4 above)
Comment (Bloodiko) [reply league_rpc_linux/lobby.py:1]: Oh, its legacy code from the first test - forgot to remove it after switching to lcu-driver. 
Going to remove it in a bit.

File: league_rpc_linux/__main__.py (hunk 8 above)
Comment (Its-Haze) [inline league_rpc_linux/__main__.py:47]: also, please name it something more descriptive than "p"

File: league_rpc_linux/__main__.py [hunk 9]
```diff
@@ -116,18 +128,15 @@ def main(cli_args: argparse.Namespace):
                             time.sleep(10)
 
                 case "InLobby":
```
Comment (Its-Haze) [inline league_rpc_linux/__main__.py:130]: Since there is now no code in here. how will we know what is happening when this case statement is executed.

Right now, it's just a sleep. Do we want to keep that in?
Should lcu driver perhaps only be executed when we reach this case statement to avoid that process to be running in the background while players are in game?

File: league_rpc_linux/__main__.py [hunk 10]
```diff
@@ -141,6 +150,9 @@ def main(cli_args: argparse.Namespace):
 
 
 if __name__ == "__main__":
+    # Patch for asyncio - read more here: https://pypi.org/project/nest-asyncio/
+    nest_asyncio.apply()
```
Comment (Its-Haze) [inline league_rpc_linux/__main__.py:154]: Very nice

File: league_rpc_linux/const.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
+# Discord Application: League of Linux
+DEFAULT_CLIENT_ID = "1185274747836174377"
+DISCORD_PROCESS_NAMES = ["Discord", "DiscordPTB", "DiscordCanary", "electron"]
+LEAGUE_OF_LEGENDS_LOGO = "https://github.com/Its-Haze/league-rpc-linux/blob/master/assets/leagueoflegends.png?raw=true"
+SMALL_TEXT = "github.com/Its-Haze/league-rpc-linux"
+CURRENT_PATCH = "13.24.1"
```
Comment (Its-Haze) [inline league_rpc_linux/const.py:6]: Do we still need CURRENT_PATCH??

I removed it from my latest release because it was no longer needed anymore.

File: league_rpc_linux/const.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+
+ALL_GAME_DATA_URL = "https://127.0.0.1:2999/liveclientdata/allgamedata"
+
+BASE_SKIN_URL = "https://ddragon.leagueoflegends.com/cdn/img/champion/tiles/"
+BASE_CHAMPION_URL = (
```
Comment (Its-Haze) [inline league_rpc_linux/const.py:12]: This has been changed in the latest release. please take a look.

File: league_rpc_linux/__main__.py (hunk 8 above)
Comment (Bloodiko) [reply league_rpc_linux/__main__.py:47]:
```python
    # Start LCU_Thread
    # This process will connect to the LCU API and updates the rpc based on data subscribed from the LCU API.
    # In this case passing the rpc object to the process is easier than trying to return updated data from the process.
    # Every In-Client update will be handled by the LCU_Thread process and will update the rpc accordingly.
```

Comment (Bloodiko) [reply league_rpc_linux/__main__.py:47]: renamed to lcu_process

File: league_rpc_linux/const.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    f"http://ddragon.leagueoflegends.com/cdn/{CURRENT_PATCH}/img/champion/"
+)
+BASE_MAP_URL = f"http://ddragon.leagueoflegends.com/cdn/{CURRENT_PATCH}/img/map/map"
```
Comment (Its-Haze) [inline league_rpc_linux/const.py:15]: If we upload the map image to "assets" folder. and point to the github url.. then we don't have to specify the current_patch. Since the map image should never change.

But i know u said you had found the map from community dragon so i will wait until you update the PR with that.

File: league_rpc_linux/lobby.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+lcu = getLcuAPI()
+
+
+def checkLcuAPI():
+    print(lcu.getDataFromPath("/lol-summoner/v1/current-summoner").items())
+
+
+def inClient():
```
Comment (Its-Haze) [inline league_rpc_linux/lobby.py:10]: Always add type hints. in parameters, and in return types.

File: league_rpc_linux/__main__.py (hunk 9 above)
Comment (Bloodiko) [reply league_rpc_linux/__main__.py:130]: I think we can just keep it running in the background as the WS will not get any updates (at least not for our purposes) and thus will not update RPC. 
There is no mesurable perfomance decrease. 
The Process Idles unless there is a data update (to the data which we subscribed to) from the WS. - No CPU time is used at all.
Everything else we dont subscribe to is not even sent by the API, it never gets passed through the ws, and doesnt reach the process - which definitely is super nice.

File: .gitignore (hunk 3 above)
Comment (Its-Haze) [reply .gitignore:26]: league-rpc-env makes sense to add here, sure. 

Not sure what getConnectDetails.py has inside of it.. maybe it's something useful during development and doesn't need to be ignored by git?

File: league_rpc_linux/processes/LCU_Thread.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    )
+    lobbyQueueInfo = await lobbyQueueInfoRaw.json()
+    data["queue"] = lobbyQueueInfo["name"]
+    data["max_players"] = lobbyQueueInfo["maximumParticipantListSize"]
+    data["map_id"] = lobbyQueueInfo["mapId"]
+    data["gamemode"] = lobbyQueueInfo["gameMode"]
+    data["lobbyQueueIsRanked"] = lobbyQueueInfo["isRanked"]
+
+
+async def update_rpc():
```
Comment (Its-Haze) [inline league_rpc_linux/processes/LCU_Thread.py:166]: Why don't we update rpc, inside of the websocket functions?

**Comment by Its-Haze** [comment]
Bug 1

Custom games & Practice tool does not change RPC.

## Steps to reproduce
1. Start league-rpc-linux
2. Create a Custom lobby / practice tool
3. Observe your discord presence.

## Expected result
rpc showing that you are creating a custom/practice tool game.

## Actual result
rpc not updating.

File: .gitignore (hunk 3 above)
Comment (Bloodiko) [reply .gitignore:26]:
```python
import psutil


def get_process_command(process_name: str) -> str:
    """
    Get process command of a process.
    """
    for proc in psutil.process_iter():
        try:
            if process_name.lower() in proc.name().lower():
                return proc.cmdline()
        except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.ZombieProcess):
            pass
    return False


def get_lcu_connect_details(proc_name: str) -> bool:
    """
    Get LCU connect details from command.
    """
    cmd = get_process_command(proc_name)
    if not cmd:
        return False
    for arg in cmd:
        if "--remoting-auth-token=" in arg:
            print(arg.split("=")[1])
        elif "--app-port=" in arg:
            print(arg.split("=")[1])
    return True

get_lcu_connect_details(proc_name="CrBrowserMain")

```

Just a small simple prog to grab the cmdline from the process and print it.

File: league_rpc_linux/__main__.py (hunk 7 above)
Comment (Bloodiko) [reply league_rpc_linux/__main__.py]: changed to import only the necessary function

File: league_rpc_linux/__main__.py (hunk 6 above)
Comment (Bloodiko) [reply league_rpc_linux/__main__.py:4]: I'm not really too familiar with either one, for me it seemed more logical to use multiprocessing, as its logically running separated from the main with the websocket. 
I guess threading wouldve been also a viable option, thats beyond my knowledge though. 
So not really a thoughtful decision, but a working one..

File: league_rpc_linux/const.py (hunk 11 above)
Comment (Bloodiko) [reply league_rpc_linux/const.py]: Removed.

File: league_rpc_linux/const.py (hunk 12 above)
Comment (Bloodiko) [reply league_rpc_linux/const.py]: Adjusted.

File: league_rpc_linux/const.py (hunk 13 above)
Comment (Bloodiko) [reply league_rpc_linux/const.py]: Adjusted to cdragon

File: league_rpc_linux/lobby.py (hunk 14 above)
Comment (Bloodiko) [reply league_rpc_linux/lobby.py]: invalid, as not existing anymore.

File: league_rpc_linux/processes/LCU_Thread.py
Comment (Bloodiko) [reply league_rpc_linux/processes/LCU_Thread.py:1]: adjusted to lower case

File: league_rpc_linux/processes/LCU_Thread.py (hunk 15 above)
Comment (Bloodiko) [reply league_rpc_linux/processes/LCU_Thread.py]: as in some cases different data is needed. - also this would sometimes override the rpc info with data we dont want overridden. 

Like being in champ select and the final "DELETE" from the queue is executed, it would remove the queue info and display the basic "idle in client" again, because it was the last "update" we got. - Until the next ChampSelect update of course. 

It also allows for "rate-limiting" RPC updates. - some events get executed mutliple times a second (Opening a lobby triggers 4 or 5 lobby updates, for which each of those would trigger rpc updates), we dont want to spam the RPC. 

In my current state, i've rate limited the rpc updates to 1 per second. - See delayUpdate()

**Comment by Bloodiko** [comment]
> Bug 1
fixed

> Bug 2
also fixed

File: league_rpc_linux/processes/lcu_thread.py [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+    "map_id": None,  # 11, 12, 21, 22, 30
+    "lobbyQueueIsRanked": False,  # True, False
+    "lobby_id": None,
+    "availability": "Idle",  # Idle, Away
+    "gameflowPhase": None,  # Lobby, Matchmaking, ReadyCheck, ChampSelect, InProgress, WaitingForStats, EndOfGame
+}
+
+## Timers ##
+
+scheduledUpdate = False
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:46]: Why is this needed?

File: league_rpc_linux/processes/lcu_thread.py [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+
+
+# As some events are called multiple times, we should limit the amount of updates to the RPC.
+# Collect update events for 1 second and then update the RPC.
+def delayUpdate():
+    global scheduledUpdate
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:52]: Why are we using a global variable?

From my understanding, global variables should be avoided at all cost.

File: league_rpc_linux/processes/lcu_thread.py [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+## Timers ##
+
+scheduledUpdate = False
+
+
+# As some events are called multiple times, we should limit the amount of updates to the RPC.
+# Collect update events for 1 second and then update the RPC.
+def delayUpdate():
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:51]: please don't use camelCased function names. always use snake_case in python

File: league_rpc_linux/processes/lcu_thread.py (hunk 16 above)
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:46]: same with variable names.. they should not be camelCased.

File: league_rpc_linux/processes/lcu_thread.py [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+    global scheduledUpdate
+
+    if not scheduledUpdate:
+        scheduledUpdate = True
+        Timer(1.0, update_rpc_and_reset_flag).start()
+
+
+def update_rpc_and_reset_flag():
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:59]: I'm having a hard time understanding why we need this. what does this achieve?

File: league_rpc_linux/processes/lcu_thread.py [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+
+    delayUpdate()
+
+
+# ranked stats
+
+
+## queue
+@connector.ws.register("/lol-matchmaking/v1/search", event_types=("CREATE",))
+async def matchmaking(connection, event):
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:192]: Since this function name is defined twice, could we change it to better fit the purpose of the function?

File: league_rpc_linux/processes/lcu_thread.py [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+        # In Queue
+        rpc.update(
+            large_image=BASE_MAP_ICON_URL.format(
+                map_name=MAP_ICON_CONVERT_MAP.get(data["map_id"])
+            ),
+            large_text=f"{GAME_MODE_CONVERT_MAP.get(data['gamemode'], data['gamemode'])}",
+            small_image=LEAGUE_OF_LEGENDS_LOGO,
+            small_text="League of Legends Logo",
+            details=f"In Queue: {data['queue']}",
+            state=f"Searching for Game...",
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:316]: Sine no variables are used inside, we don't have to turn this string into an "f string"

File: league_rpc_linux/processes/lcu_thread.py [hunk 22]
```diff
@@ -0,0 +1,355 @@
...
+        # In Champ Select
+        rpc.update(
+            large_image=BASE_MAP_ICON_URL.format(
+                map_name=MAP_ICON_CONVERT_MAP.get(data["map_id"])
+            ),
+            large_text=f"{GAME_MODE_CONVERT_MAP.get(data['gamemode'], data['gamemode'])}",
+            small_image=LEAGUE_OF_LEGENDS_LOGO,
+            small_text="League of Legends Logo",
+            details=f"In Champ Select: {data['queue']}",
+            state=f"Picking Champions...",
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:301]: Using an f-string that does not have any interpolated variables

**Comment by Its-Haze** [comment]
I tested it locally now and honestly, good job! It feels good and seems to work as well.

However there are a few key points i would like to be fixed.
- Format the code using Black, and sort the imports using Isort.
- Rename variables, function names to be snake_cased.. not camelCased. This is python and we should follow the naming conventions for that.
- the small_image of the rpc, should have the text "github.com/Its-Haze/league-rpc-linux" just as it does in game.
- I honestly think the summoner icon looks so fucking good, that i believe it should be the icon in all instances.. meaning that the summoner icon should be the large_image in rpc, even if the user is in a lobby/queue/champ select..
And then the small_image should maybe change between being in lobby/inqueue/inchamp select.. but i am not sure what icons to pick there yet..
- When the user is in the client (not in lobby) then rpc status will show "idle / away".. instead have it say "Online / Away".
- Please add type hints on all functions, parameters, and return types.

---
Do you think it makes sense to have a command line argument for "--no-icon or --no-summoners-icon" to let people opt out of showing their ingame summoners icon.. (It should then default back to the original league images that we have)?

File: league_rpc_linux/processes/lcu_thread.py [hunk 23]
```diff
@@ -1,355 +1,370 @@
...
-
-    delayUpdate()
-
-
-@connector.ws.register("/lol-gameflow/v1/gameflow-phase", event_types=("UPDATE",))
-async def gameflow_phase_updated(connection, event):
-    print("Gameflow Phase has been updated.")
-    data["gameflowPhase"] = event.data  # returns plain string of the phase
+class LcuThread:
+    self_obj: None
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:39]: What's this for?

File: league_rpc_linux/processes/lcu_thread.py [hunk 24]
```diff
@@ -1,355 +1,370 @@
...
+
+    def __init__(self, rpc_from_main):
+        self.rpc = rpc_from_main
+        self.connector = lcu_driver.Connector()
+        LcuThread.connector = self.connector
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:44]: I don't think this line is nessesary

File: league_rpc_linux/processes/lcu_thread.py [hunk 25]
```diff
@@ -1,355 +1,370 @@
...
+            "queue_id": -1,
+            "map_id": None,  # 11, 12, 21, 22, 30
+            "is_ranked_queue": False,  # True, False
+            "lobby_id": None,
+            "availability": "Idle",  # Idle, Away
+            "gamflow_phase": None,  # Lobby, Matchmaking, ReadyCheck, ChampSelect, InProgress, WaitingForStats, EndOfGame
+        }
+
+        self.register_ws_events(self.connector)
+        print("Starting LCU API.")
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:74]: does it really start here in the __init__, or in the lcu_thread.connector.start() call?

File: league_rpc_linux/processes/lcu_thread.py [hunk 26]
```diff
@@ -1,355 +1,370 @@
...
+
+    @staticmethod
+    def start_connector(rpc_from_main):
+        lcu_thread = LcuThread(rpc_from_main)
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:78]: You can use a classmethod, instead of a staticmethod, if you want to construct the class from within a method. But I'm not really sure why this would be needed

File: league_rpc_linux/processes/lcu_thread.py [hunk 27]
```diff
@@ -1,355 +1,370 @@
...
+        lcu_thread.connector.start()
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:79]: from the def __init__ you have already called Connector() and stored it in an instance variable ""self.connector"

You could continue to use that instance variable by using "self.connector.start()" instead of creating an entire new LcuThread i believe

File: league_rpc_linux/processes/lcu_thread.py [hunk 28]
```diff
@@ -1,355 +1,370 @@
...
+        LcuThread.self_obj = lcu_thread
+
+    # Register WS Events
+    # The Decorator will not work inside classes, thus we need to explicitly register the events manually.
+    def register_ws_events(self, connector):
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:84]: pretty cool, pretty cool!

File: league_rpc_linux/processes/lcu_thread.py [hunk 29]
```diff
@@ -1,355 +1,370 @@
...
+        self.data["level"] = event.data["summonerLevel"]
+        self.data["summoner_id"] = event.data["summonerId"]
+        self.data["summoner_icon"] = event.data["profileIconId"]
+
+        self.update_scheduler.delay_update()
+
+    async def chat_updated(self, connection, event):
+        print("Chat has been updated.")
+        if event.data["availability"] == "chat":
+            self.data["availability"] = "Idle"
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:137]: please change it to "Online"

File: league_rpc_linux/processes/lcu_thread.py [hunk 30]
```diff
@@ -1,355 +1,370 @@
...
-            large_image=BASE_MAP_ICON_URL.format(
-                map_name=MAP_ICON_CONVERT_MAP.get(data["map_id"])
-            ),
-            large_text=f"{GAME_MODE_CONVERT_MAP.get(data['gamemode'], data['gamemode'])}",
-            small_image=LEAGUE_OF_LEGENDS_LOGO,
-            small_text="League of Legends Logo",
-            details=f"In Champ Select: {data['queue']}",
-            state=f"Picking Champions...",
-            party_size=[data["players"], data["max_players"]],
+        summonerData = await summonerDataRaw.json()
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:240]: Do we have a reason for separating summonerDataRaw, and summonerData?

Why not just slap ``.json()``  at the end of ``await connection.request(
            "GET", "/lol-summoner/v1/current-summoner"
        )``
        
 to get ``summoner_data = await connection.request(
            "GET", "/lol-summoner/v1/current-summoner"
        ).json()``

Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py:240]: I actually tried that. But it didn't yield any results, as if it loaded forever and got stuck somewhere in a loop

File: league_rpc_linux/processes/lcu_thread.py (hunk 27 above)
Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py]: changed connector.start inside the init

File: league_rpc_linux/processes/lcu_thread.py (hunk 26 above)
Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py]: moved out of class, only init is invoked. no other operation is done.

File: league_rpc_linux/processes/lcu_thread.py (hunk 24 above)
Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py]: removed legacy code

File: league_rpc_linux/processes/lcu_thread.py (hunk 22 above)
Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py]: rm unnecessary f string usage

File: league_rpc_linux/processes/lcu_thread.py (hunk 20 above)
Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py]: consolidated into a single funtion

File: league_rpc_linux/processes/lcu_thread.py (hunk 19 above)
Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py]: I dont want to spam RPC with multiple updates per second in case the websocket update functions are triggered multiple times. 
Like /lobby is triggered really often in short succession ~1sec = 4-5 calls.

**Comment by Bloodiko** [comment]
> Do you think it makes sense to have a command line argument for "--no-icon or --no-summoners-icon" to let people opt out of showing their ingame summoners icon.. (It should then default back to the original league images that we have)?

I'd suggest an config file. 
Then when running with command line arg --set config.stuff=value  or similar - this will require an additional help page, and some more work though. 


I've implemented your comments. - Idle --> Online,  typehints, ... 
Issue is, i'm unable to get rid of the Error when using the class approach. 
it worked when not using classes. - Not sure where the issue exactly is. PYTHONDEBUGASYNCIO=1  is also really vage about whats happening. 
Sometimes its in the rpc library - sometimes inside the lcu_driver library. 

Maybe its a weird interaction between nest_asyncio and self inside classes ? - this whole asyncio lib feels somehow unfinished. you need to await any async function inside an async function unless you schedule it with asyncio.run yourself, but thats not what we need. - is that really the purpose of async ? - that seems weird. 


(leaving the error here, for google search references eventually. ) 
```bash
Exception in callback _SelectorSocketTransport._read_ready()
handle: <Handle _SelectorSocketTransport._read_ready()>
Traceback (most recent call last):
  File "/usr/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
RuntimeError: cannot enter context: <_contextvars.Context object at 0x7f16400229c0> is already entered
```

File: league_rpc_linux/processes/lcu_thread.py (hunk 30 above)
Comment (Its-Haze) [reply league_rpc_linux/processes/lcu_thread.py:264]: that's strange. alright then

**Comment by Bloodiko** [comment]
I'd suggest rolling back to the non-classes approach. 
And see from there. 
As this does not work at all. Tons of error I don't know where they come from.

**Comment by Its-Haze** [comment]
Yea i am not an expert in asyncio or event loops but something does not seem to go hand in hand together.. Not sure if that is because of how we use it. Or if we are going with the wrong approach..
If you believe we should stick with the old version (without classes) then let's do that. We have tried classes and it clearly gives more problems than it was expected to solve.

**Comment by Its-Haze** [comment]
I can help out with type hinting. and some of the other details i commented on

Please rollback/revert the class change and go back to the commit where it was working as expected. and let's continue from there.

File: league_rpc_linux/processes/lcu_thread.py [hunk 31]
```diff
@@ -1,5 +1,6 @@
 import time
 from threading import Timer
+from dataclasses import dataclass, field
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:3]: Why is field imported here?

can't see it being used.

Comment (Bloodiko) [reply league_rpc_linux/processes/lcu_thread.py:3]: Oh right , had a test with it and forgot to remove it. 
Done 👍

File: league_rpc_linux/processes/lcu_thread.py [hunk 32]
```diff
@@ -0,0 +1,377 @@
...
+    MAP_ICON_CONVERT_MAP,
+    PROFILE_ICON_BASE_URL,
+    SMALL_TEXT,
+)
+
+
+@dataclass
+class ClientData:
+    availability: str = "Online"  # "Online", "Away
+    gamemode: str = None
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:24]: Incompatible type hinting.

You have type hinted this value as to be a string, but you are assigning a default value of "None".

The correct type hint in this case would be ``gamemode: Optional[str] = None``

where ``Optional`` is imported from ``typing`` module.

Alternatively what you can do is to set a falsey value..

so ``gamemode: str = ""`` This will work.. however you need to take that into account if you have used ``if clientdata.gamemode is not None`` somewhere.. which i don't think you have.

Comment (Its-Haze) [reply league_rpc_linux/processes/lcu_thread.py:24]: then again.. this comment applies to all variables you have typehinted as a specific type, but used "None" as it's default.

File: league_rpc_linux/processes/lcu_thread.py [hunk 33]
```diff
@@ -0,0 +1,377 @@
...
+    def __init__(self):
+        self.scheduled_update = False
+
+    def delay_update(self):
+        if not self.scheduled_update:
+            self.scheduled_update = True
+            Timer(1.0, self.update_rpc_and_reset_flag).start()
+
+    def update_rpc_and_reset_flag(self):
+        module_data.update_rpc()  # Assuming update_rpc is defined elsewhere
```
Comment (Its-Haze) [inline league_rpc_linux/processes/lcu_thread.py:57]: the type of module_data is unknown. because the class RPCUpdater has no idea if an instance, called "module_data" is anywhere in this scope. 

Consider sending in the module_data as a parameter for "update_rpc_and_reset_flag" and type hint it correctly, so that it is aware of the function it's about to run.

The easiest way would be to split all of the different functionalities in this file, into separate modules and then import them all here.. we don't have to have the update_rpc inside of a class. i think.

I will test it myself and see if i can get something done, and perhaps send a PR to this branch

Comment (Its-Haze) [reply league_rpc_linux/processes/lcu_thread.py]: Please check latest commit. This comment has been fixed.

File: league_rpc_linux/processes/lcu_thread.py (hunk 32 above)
Comment (Its-Haze) [reply league_rpc_linux/processes/lcu_thread.py]: This has been fixed in the latest commit.

**Review by Its-Haze — APPROVED** [review]

**Comment by Its-Haze** [comment]
Thank you for your pull request @Bloodiko !
